### PR TITLE
fix(trait):make the default trait name not the same as the component name

### LINKED ISSUE #66: Make the default trait name not the same as the component name
Currently, we use the component name as the trait name in the `setTraitProperties` function. It will be good for the name to indicate that it belongs to a `trait`

### PR DESCRIPTION
**Pull request by Ghostbaby**
Make the default trait name not the same as the component name

Fixes #66

Signed-off-by: zhuhuijun <zhuhuijunzhj@gmail.com>

### CODE REVIEW

**Comment by wonderflow** [comment]
in Kubernetes, children resources have the name of their parents as prefix like:

* Deployment: deploymentname
* ReplicaSet: deploymentname-dsjn3ndshh
* Pod: deploymentname-dsjn3ndshh-idsjn

I suggest we do something like it here.

* ComponentName: my-comp
* TraitName: my-comp-trait-dsjn3ndshh

File: pkg/oam/util/helper.go [hunk 1]
```diff
@@ -32,6 +32,11 @@ var (
 	ReconcileWaitResult = reconcile.Result{RequeueAfter: 30 * time.Second}
 )
 
+const (
+	//TraitPrefixKey is prefix of trait name
+	TraitPrefixKey = "trait"
```
Comment (ryanzhang-oss) [inline pkg/oam/util/helper.go:37]: I don't seem to see this being used anywhere else, why make it public?

File: pkg/oam/util/helper.go [hunk 2]
```diff
@@ -197,3 +202,8 @@ func Object2Map(obj interface{}) (map[string]interface{}, error) {
 	err = json.Unmarshal(bts, &res)
 	return res, err
 }
+
+// GenTraitName generate trait name
```
Comment (ryanzhang-oss) [inline pkg/oam/util/helper.go:206]: please add a unit test for this function

File: pkg/oam/util/helper.go [hunk 3]
```diff
@@ -197,3 +202,8 @@ func Object2Map(obj interface{}) (map[string]interface{}, error) {
...
+func GenTraitName(componentName string) string {
+	return fmt.Sprintf("%s-%s", TraitPrefixKey, componentName)
```
Comment (ryanzhang-oss) [inline pkg/oam/util/helper.go:208]: I think @wonderflow has a good suggestion on the naming convention

File: pkg/oam/util/helper.go (hunk 2 above)
Comment (Ghostbaby) [reply pkg/oam/util/helper.go:206]: ok

File: pkg/oam/util/helper.go (hunk 1 above)
Comment (Ghostbaby) [reply pkg/oam/util/helper.go:37]: It will be used in `GenTraitName`.

File: pkg/oam/util/helper.go (hunk 3 above)
Comment (Ghostbaby) [reply pkg/oam/util/helper.go]: ok

File: pkg/controller/v1alpha2/applicationconfiguration/render.go [hunk 4]
```diff
@@ -165,12 +165,13 @@ func (r *components) renderComponent(ctx context.Context, acc v1alpha2.Applicati
 }
 
 func (r *components) renderTrait(ctx context.Context, ct v1alpha2.ComponentTrait, namespace, componentName string, ref *metav1.OwnerReference, dag *dependency.DAG) (*unstructured.Unstructured, *v1alpha2.TraitDefinition, error) {
+	traitName := util.GenTraitName(componentName, ct.DeepCopy(), nil)
```
Comment (wonderflow) [inline pkg/controller/v1alpha2/applicationconfiguration/render.go:168]: I think we must check if there's already an exist trait, before generate a new traitname

File: pkg/oam/util/helper.go (hunk 1 above)
Comment (ryanzhang-oss) [reply pkg/oam/util/helper.go:43]: GenTraitName is in the same package....

File: pkg/oam/util/helper.go [hunk 5]
```diff
@@ -197,3 +208,39 @@ func Object2Map(obj interface{}) (map[string]interface{}, error) {
...
+// a collisionCount to avoid hash collision. The hash will be safe encoded to
+// avoid bad words.
+func ComputeHash(trait *v1alpha2.ComponentTrait, collisionCount *int32) string {
+	componentTraitHasher := fnv.New32a()
+	DeepHashObject(componentTraitHasher, *trait)
+
+	// Add collisionCount in the hash if it exists.
+	if collisionCount != nil {
+		collisionCountBytes := make([]byte, 8)
+		binary.LittleEndian.PutUint32(collisionCountBytes, uint32(*collisionCount))
```
Comment (ryanzhang-oss) [inline pkg/oam/util/helper.go:227]: collisionCount is always passed in as new(int32) which I assume is a pointer to zero, I wonder why do we need this parameter?

File: pkg/controller/v1alpha2/applicationconfiguration/render.go (hunk 4 above)
Comment (Ghostbaby) [reply pkg/controller/v1alpha2/applicationconfiguration/render.go]: done

File: pkg/oam/util/helper.go (hunk 5 above)
Comment (Ghostbaby) [reply pkg/oam/util/helper.go]: yes, I will remove this parameter.

**Comment by zzxwill** [comment]
> in Kubernetes, children resources have the name of their parents as prefix like:
> 
> * Deployment: `deploymentname`
> * ReplicaSet: `deploymentname-dsjn3ndshh`
> * Pod: `deploymentname-dsjn3ndshh-idsjn`
> 
> I suggest we do something like it here.
> 
> * AppConfig: `appconfigname` // unique globally
> * Component: `componentname`  // unique globally
> * Trait: `componentname-trait-dsjn3ndshh` // generate a hash suffix
> 
> If we don't assign name, we can say there's only one instance for one kind of trait. So we can get from ownerRef before create new.

`unique globally` means cross namespaces? Or it will allow same names in different names?

File: pkg/oam/util/helper_test.go [hunk 6]
```diff
@@ -292,3 +292,41 @@ var _ = Describe("Test unstructured related helper utils", func() {
...
+			},
+		}
+		for _, test := range test {
+			name := fmt.Sprintf("%s-%s-%s", test.name, util.TraitPrefixKey, util.ComputeHash(test.template))
+			traitName := util.GenTraitName(test.name, test.template)
+			Expect(traitName).Should(Equal(name))
+
+			new := test.template.DeepCopy()
+			new.Trait = runtime.RawExtension{
+				Object: &mts,
```
Comment (ryanzhang-oss) [inline pkg/oam/util/helper_test.go:327]: what does this test?

File: pkg/oam/util/helper_test.go [hunk 7]
```diff
@@ -292,3 +292,41 @@ var _ = Describe("Test unstructured related helper utils", func() {
...
+			Spec: v1alpha2.ManualScalerTraitSpec{
+				ReplicaCount: 3,
+			},
+		}
+
+		test := []struct {
+			name                string
+			template            *v1alpha2.ComponentTrait
+			collisionCount      *int32
+			otherCollisionCount *int32
```
Comment (ryanzhang-oss) [inline pkg/oam/util/helper_test.go:313]: I don't seem to see those being used

File: pkg/oam/util/helper_test.go [hunk 8]
```diff
@@ -292,3 +292,41 @@ var _ = Describe("Test unstructured related helper utils", func() {
 		}
 	})
 })
+
+var _ = Describe("Test GenTraitName helper utils", func() {
+	It("Test generate trait name", func() {
```
Comment (ryanzhang-oss) [inline pkg/oam/util/helper_test.go:297]: This test actually relies on the ComputeHash function. I wonder if we can have an "expected" value that is not computed by the ComputeHash function?

File: pkg/controller/v1alpha2/applicationconfiguration/render.go [hunk 9]
```diff
@@ -397,3 +403,39 @@ func addDataInputsToDAG(dag *dependency.DAG, ins []v1alpha2.DataInput, obj *unst
...
+
+		for _, cr := range crs.Items {
+			for _, owner := range cr.GetOwnerReferences() {
+				if owner.UID == ref.UID {
+					traitName = cr.GetName()
+				}
+			}
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (Ghostbaby) [reply pkg/controller/v1alpha2/applicationconfiguration/render.go:435]: Maybe we can store traitName to annotations.

File: pkg/oam/util/helper_test.go (hunk 7 above)
Comment (Ghostbaby) [reply pkg/oam/util/helper_test.go:313]: I will remove these parameters.

File: pkg/oam/util/helper_test.go (hunk 6 above)
Comment (Ghostbaby) [reply pkg/oam/util/helper_test.go:327]: The main test is to pass in the same `ComponentTrait` to get the same hash value.

File: pkg/oam/util/helper_test.go (hunk 8 above)
Comment (Ghostbaby) [reply pkg/oam/util/helper_test.go:297]: done

File: pkg/oam/util/helper_test.go [hunk 10]
```diff
@@ -292,3 +292,46 @@ var _ = Describe("Test unstructured related helper utils", func() {
...
+			newTemplate := test.template.DeepCopy()
+			newTemplate.Trait = runtime.RawExtension{
+				Object: &mts,
+			}
+			newTraitName := util.GenTraitName(test.name, newTemplate)
+			Expect(oldTraitName).ShouldNot(Equal(newTraitName))
+
+			newTemplate.Trait = runtime.RawExtension{}
+
+			lastedTraitName := util.GenTraitName(test.name, newTemplate)
```
Comment (ryanzhang-oss) [inline pkg/oam/util/helper_test.go:332]: nit: lastTraitName ?

Comment (ryanzhang-oss) [inline pkg/oam/util/helper_test.go:332]: a table style test should not have logics in the last loop, instead, they should be test cases. You can move those to test cases and add a result (should equal or not equal) there.

File: pkg/controller/v1alpha2/applicationconfiguration/render.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ryanzhang-oss) [reply pkg/controller/v1alpha2/applicationconfiguration/render.go:464]: I think the appConfig CR actually has a status field that you can store the trait name

File: pkg/oam/util/helper_test.go (hunk 10 above)
Comment (Ghostbaby) [reply pkg/oam/util/helper_test.go:332]: done

File: pkg/controller/v1alpha2/applicationconfiguration/render.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (Ghostbaby) [reply pkg/controller/v1alpha2/applicationconfiguration/render.go:464]: ok, I will get the history trait name form the status field.

**Comment by Ghostbaby** [comment]
@ryanzhang-oss ok, I will get the history trait name form the status field.

File: pkg/controller/v1alpha2/applicationconfiguration/render_test.go [hunk 11]
```diff
@@ -907,6 +912,79 @@ func TestGetComponent(t *testing.T) {
 	}, c)
 }
 
+var scheme = runtime.NewScheme()
+
+func TestRenderTraitName(t *testing.T) {
```
Comment (ryanzhang-oss) [inline pkg/controller/v1alpha2/applicationconfiguration/render_test.go:917]: IIRC, this is the case that a trait name exist. It seems that it misses the case that trait doesn't exist

File: pkg/controller/v1alpha2/applicationconfiguration/render.go [hunk 12]
```diff
@@ -426,3 +429,33 @@ func addDataInputsToDAG(dag *dependency.DAG, ins []v1alpha2.DataInput, obj *unst
 		dag.AddSink(in.ValueFrom.DataOutputName, obj, attaches, in.ToFieldPaths)
 	}
 }
+
+// GetTraitName return trait name
+func getTraitName(ac *v1alpha2.ApplicationConfiguration, componentName string,
+	ct *v1alpha2.ComponentTrait, t *unstructured.Unstructured) string {
+	var (
```
Comment (wonderflow) [inline pkg/controller/v1alpha2/applicationconfiguration/render.go:436]: I assume we can check `t.GetName()` first, if this is not null, just return it. That's also the traitName specified by user.

File: pkg/controller/v1alpha2/applicationconfiguration/render_test.go (hunk 11 above)
Comment (Ghostbaby) [reply pkg/controller/v1alpha2/applicationconfiguration/render_test.go:917]: done

File: pkg/controller/v1alpha2/applicationconfiguration/render.go (hunk 12 above)
Comment (Ghostbaby) [reply pkg/controller/v1alpha2/applicationconfiguration/render.go:581]: done

**Comment by Ghostbaby** [comment]
@wonderflow @zzxwill @ryanzhang-oss @sunny0826  Please help me review this PR. At the same time, the CI UT coverage calculation seems to have problems.

**Comment by ryanzhang-oss** [comment]
@Ghostbaby  please kick the build process again, it seems to be stuck

**Pull request merged**